### PR TITLE
exclude pattern documentation

### LINKED ISSUE #356: exclude not working
phpmd 2.4.2

command line exclude not working such as;

--exclude="/home/jjpet/svnWork/225161/includes/config/" and variations without the = " and trailing /

### PR DESCRIPTION
**Pull request by sveneld**
exclude pattern unittest

Type: documentation update + unit test  
Issue: Resolves #356 #373 
Breaking change: no

### CODE REVIEW

**Comment by sveneld** [comment]
@ravage84

File: README.rst [hunk 1]
```diff
@@ -123,7 +123,7 @@ Command line options
     extensions, e.g. php,phtml.
 
   - ``--exclude`` - Comma-separated string of patterns that are used to ignore
-    directories.
+    directories. Use PCRE pattern for preg_match. For example `*src/foo/*.php` or `*src/foo/*`
```
Comment (kylekatarnls) [inline README.rst:126]: That's not exact. We do:
```php
$quoted = array_map('preg_quote', $patterns);
$this->relative = '(' . str_replace('\*', '.*', join('|', $quoted)) . ')i';
```

That means we actually escape PCRE characters from the config, then we replace `"*"` input with `".*"` which allows one or more characters in RegExps.

We will actually not handle properly PCRE patterns (if you use `.*` it won't work as expected, as for any `[0-9]` etc.) so I would simply say:

> Use asterisks to exclude by pattern. For example `*src/foo/*.php` or `*src/foo/*`

File: src/main/php/PHPMD/TextUI/CommandLineOptions.php [hunk 2]
```diff
@@ -102,6 +102,7 @@ class CommandLineOptions
 
     /**
      * A string of comma-separated pattern that is used to exclude directories.
+     * Use PCRE pattern for preg_match. For example *src/foo/*.php or *src/foo/*
```
Comment (kylekatarnls) [inline src/main/php/PHPMD/TextUI/CommandLineOptions.php:105]: Same here.

File: src/main/php/PHPMD/PHPMD.php [hunk 3]
```diff
@@ -175,7 +175,7 @@ public function processFiles(
     ) {
 
         // Merge parsed excludes
-        $this->ignorePatterns = array_merge($this->ignorePatterns, $ruleSetFactory->getIgnorePattern($ruleSets));
+        $this->setIgnorePattern($ruleSetFactory->getIgnorePattern($ruleSets));
```
Comment (tvbeek) [inline src/main/php/PHPMD/PHPMD.php:178]: Detail for others, this looks like a BC break but the setIgnore doesn't set but merge the patterns to the exisitng ones.

File: README.rst (hunk 1 above)
Comment (sveneld) [reply README.rst]: ++

File: src/main/php/PHPMD/TextUI/CommandLineOptions.php (hunk 2 above)
Comment (sveneld) [reply src/main/php/PHPMD/TextUI/CommandLineOptions.php]: ++

File: src/main/php/PHPMD/PHPMD.php (hunk 3 above)
Comment (ravage84) [reply src/main/php/PHPMD/PHPMD.php:178]: I think we should address this confusion. Either by introducing a new, correctly named method and deprecating the old or by other means. /cc  @kylekatarnls

Comment (kylekatarnls) [reply src/main/php/PHPMD/PHPMD.php:178]: Note that also both the setter and the getter are also singular while they should be plural as handling multiple patterns.

File: README.rst (hunk 1 above)
Comment (ravage84) [reply README.rst]: Was this adddressed?

Comment (kylekatarnls) [reply README.rst]: Yes, the new description is correct.

**Pull request merged**